File: extractor/src/test/java/org/schabi/newpipe/extractor/services/bandcamp/BandcampPlaylistExtractorTest.java
// Created by Fynn Godau 2019, licensed GNU GPL version 3 or later

package org.schabi.newpipe.extractor.services.bandcamp;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;
import org.schabi.newpipe.downloader.DownloaderTestImpl;
import org.schabi.newpipe.extractor.NewPipe;
import org.schabi.newpipe.extractor.exceptions.ContentNotAvailableException;
import org.schabi.newpipe.extractor.exceptions.ExtractionException;
import org.schabi.newpipe.extractor.exceptions.ParsingException;
import org.schabi.newpipe.extractor.playlist.PlaylistExtractor;
import org.schabi.newpipe.extractor.services.BasePlaylistExtractorTest;
import org.schabi.newpipe.extractor.services.bandcamp.extractors.BandcampPlaylistExtractor;
import org.schabi.newpipe.extractor.stream.Description;
import org.schabi.newpipe.extractor.stream.StreamInfoItem;

import java.io.IOException;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.schabi.newpipe.extractor.ExtractorAsserts.assertContains;
import static org.schabi.newpipe.extractor.ExtractorAsserts.assertContainsOnlyEquivalentImages;
import static org.schabi.newpipe.extractor.ExtractorAsserts.assertEmpty;
import static org.schabi.newpipe.extractor.ExtractorAsserts.assertNotOnlyContainsEquivalentImages;
import static org.schabi.newpipe.extractor.ServiceList.Bandcamp;

/**
 * Tests for {@link BandcampPlaylistExtractor}
 */
public class BandcampPlaylistExtractorTest {

    @BeforeAll
    public static void setUp() {
        NewPipe.init(DownloaderTestImpl.getInstance());
    }

    /**
     * Test whether playlists contain the correct amount of items
     */
    @Test
    void testCount() throws ExtractionException, IOException {
        final PlaylistExtractor extractor = Bandcamp.getPlaylistExtractor("https://macbenson.bandcamp.com/album/coming-of-age");
        extractor.fetchPage();

        assertEquals(5, extractor.streamCount);
    }

    /**
     * Tests whether different stream thumbnails (track covers) get loaded correctly
     */
    @Test
    void testDifferentTrackCovers() throws ExtractionException, IOException {
        final PlaylistExtractor extractor = Bandcamp.getPlaylistExtractor("https://zachbensonarchive.bandcamp.com/album/results-of-boredom");
        extractor.fetchPage();

        final List<StreamInfoItem> l = extractor.initialPage.getItems();
        assertContainsOnlyEquivalentImages(extractor.getThumbnails(), l.get(0).thumbnails);
        assertNotOnlyContainsEquivalentImages(extractor.getThumbnails(), l.get(5).thumbnails);
    }

    /**
     * Tests that no attempt to load every track's cover individually is made
     */
    @Test
    @Timeout(10)
    void testDifferentTrackCoversDuration() throws ExtractionException, IOException {
        final PlaylistExtractor extractor = Bandcamp.getPlaylistExtractor("https://infiniteammo.bandcamp.com/album/night-in-the-woods-vol-1-at-the-end-of-everything");
        extractor.fetchPage();

        /* All tracks on this album have the same cover art, but I don't know any albums with more
         * than 10 tracks that has at least one track with a cover art different from the rest.
         */
        final List<StreamInfoItem> l = extractor.initialPage.getItems();
        assertContainsOnlyEquivalentImages(extractor.getThumbnails(), l.get(0).thumbnails);
        assertContainsOnlyEquivalentImages(extractor.getThumbnails(), l.get(5).thumbnails);
    }

    /**
     * Test playlists with locked content
     */
    @Test
    void testLockedContent() throws ExtractionException {
        final PlaylistExtractor extractor = Bandcamp.getPlaylistExtractor("https://billwurtz.bandcamp.com/album/high-enough");

        assertThrows(ContentNotAvailableException.class, extractor::fetchPage);
    }

    /**
     * Test playlist with just one track
     */
    @Test
    void testSingleStreamPlaylist() throws ExtractionException, IOException {
        final PlaylistExtractor extractor = Bandcamp.getPlaylistExtractor("https://zachjohnson1.bandcamp.com/album/endless");
        extractor.fetchPage();

        assertEquals(1, extractor.streamCount);
    }

    public static class ComingOfAge implements BasePlaylistExtractorTest {

        private static PlaylistExtractor extractor;

        @BeforeAll
        public static void setUp() throws ExtractionException, IOException {
            NewPipe.init(DownloaderTestImpl.getInstance());
            extractor = Bandcamp.getPlaylistExtractor("https://macbenson.bandcamp.com/album/coming-of-age");
            extractor.fetchPage();
        }

        @Test
        public void testThumbnails() throws ParsingException {
            BandcampTestUtils.testImages(extractor.getThumbnails());
        }

        @Test
        public void testBanners() throws ParsingException {
            assertEmpty(extractor.getBanners());
        }

        @Test
        void testUploaderUrl() throws ParsingException {
            assertTrue(extractor.uploaderUrl.contains("macbenson.bandcamp.com"));
        }

        @Test
        public void testUploaderName() throws ParsingException {
            assertEquals("mac benson", extractor.uploaderName);
        }

        @Test
        public void testUploaderAvatars() throws ParsingException {
            BandcampTestUtils.testImages(extractor.uploaderAvatars);
        }

        @Test
        public void testStreamCount() throws ParsingException {
            assertEquals(5, extractor.streamCount);
        }

        @Test
        public void testDescription() throws ParsingException {
            final Description description = extractor.description;
            assertNotEquals(Description.EMPTY_DESCRIPTION, description);
            assertContains("Artwork by Shona Radcliffe", description.content); // about
            assertContains("All tracks written, produced and recorded by Mac Benson",
                    description.content); // credits
            assertContains("all rights reserved", description.content); // license
        }

        @Test
        @Override
        public void testUploaderVerified() throws Exception {
            assertFalse(extractor.isUploaderVerified);
        }

        @Test
        void testInitialPage() throws IOException, ExtractionException {
            assertNotNull(extractor.initialPage.getItems().get(0));
        }

        @Test
        public void testServiceId() {
            assertEquals(Bandcamp.serviceId, extractor.getServiceId());
        }

        @Test
        public void testName() throws ParsingException {
            assertEquals("Coming of Age", extractor.name);
        }

        @Test
        public void testId() throws Exception {
            assertEquals("https://macbenson.bandcamp.com/album/coming-of-age", extractor.getId());
        }

        @Test
        public void testUrl() throws Exception {
            assertEquals("https://macbenson.bandcamp.com/album/coming-of-age", extractor.getUrl());
        }

        @Test
        public void testOriginalUrl() throws Exception {
            assertEquals("https://macbenson.bandcamp.com/album/coming-of-age", extractor.getOriginalUrl());
        }

        @Test
        void testNextPageUrl() throws IOException, ExtractionException {
            assertNull(extractor.getPage(extractor.initialPage.nextPage));
        }

        @Test
        public void testRelatedItems() throws Exception {
            // DefaultTests.defaultTestRelatedItems(extractor);
            // Would fail because BandcampPlaylistStreamInfoItemExtractor.getUploaderName() returns an empty String
        }

        @Test
        public void testMoreRelatedItems() throws Exception {
        }
    }
}

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/bandcamp/BandcampPlaylistExtractorTest.kt
// Created by Fynn Godau 2019, licensed GNU GPL version 3 or later
package org.schabi.newpipe.extractor.services.bandcamp

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Timeout
import org.schabi.newpipe.downloader.DownloaderTestImpl
import org.schabi.newpipe.extractor.Extractor.serviceId
import org.schabi.newpipe.extractor.ExtractorAsserts
import org.schabi.newpipe.extractor.NewPipe.init
import org.schabi.newpipe.extractor.StreamingService.getPlaylistExtractor
import org.schabi.newpipe.extractor.exceptions.ContentNotAvailableException
import org.schabi.newpipe.extractor.exceptions.ExtractionException
import org.schabi.newpipe.extractor.exceptions.ParsingException
import org.schabi.newpipe.extractor.playlist.PlaylistExtractor
import org.schabi.newpipe.extractor.services.BasePlaylistExtractorTest
import org.schabi.newpipe.extractor.stream.Description
import org.schabi.newpipe.extractor.stream.StreamInfoItem
import java.io.IOException

/**
 * Tests for [BandcampPlaylistExtractor]
 */
class BandcampPlaylistExtractorTest {
    /**
     * Test whether playlists contain the correct amount of items
     */
    @Test
    @Throws(ExtractionException::class, IOException::class)
    fun testCount() {
        val extractor: PlaylistExtractor = Bandcamp.getPlaylistExtractor("https://macbenson.bandcamp.com/album/coming-of-age")
        extractor.fetchPage()
        assertEquals(5, extractor.streamCount)
    }

    /**
     * Tests whether different stream thumbnails (track covers) get loaded correctly
     */
    @Test
    @Throws(ExtractionException::class, IOException::class)
    fun testDifferentTrackCovers() {
        val extractor: PlaylistExtractor = Bandcamp.getPlaylistExtractor("https://zachbensonarchive.bandcamp.com/album/results-of-boredom")
        extractor.fetchPage()
        val l: List<StreamInfoItem> = extractor.initialPage.getItems()
        ExtractorAsserts.assertContainsOnlyEquivalentImages(extractor.thumbnails, l[0].thumbnails)
        ExtractorAsserts.assertNotOnlyContainsEquivalentImages(extractor.thumbnails, l[5].thumbnails)
    }

    /**
     * Tests that no attempt to load every track's cover individually is made
     */
    @Test
    @Timeout(10)
    @Throws(ExtractionException::class, IOException::class)
    fun testDifferentTrackCoversDuration() {
        val extractor: PlaylistExtractor = Bandcamp.getPlaylistExtractor("https://infiniteammo.bandcamp.com/album/night-in-the-woods-vol-1-at-the-end-of-everything")
        extractor.fetchPage()

        /* All tracks on this album have the same cover art, but I don't know any albums with more
         * than 10 tracks that has at least one track with a cover art different from the rest.
         */
        val l: List<StreamInfoItem> = extractor.initialPage.getItems()
        ExtractorAsserts.assertContainsOnlyEquivalentImages(extractor.thumbnails, l[0].thumbnails)
        ExtractorAsserts.assertContainsOnlyEquivalentImages(extractor.thumbnails, l[5].thumbnails)
    }

    /**
     * Test playlists with locked content
     */
    @Test
    @Throws(ExtractionException::class)
    fun testLockedContent() {
        val extractor: PlaylistExtractor = Bandcamp.getPlaylistExtractor("https://billwurtz.bandcamp.com/album/high-enough")
        Assertions.assertThrows(ContentNotAvailableException::class.java) { extractor.fetchPage() }
    }

    /**
     * Test playlist with just one track
     */
    @Test
    @Throws(ExtractionException::class, IOException::class)
    fun testSingleStreamPlaylist() {
        val extractor: PlaylistExtractor = Bandcamp.getPlaylistExtractor("https://zachjohnson1.bandcamp.com/album/endless")
        extractor.fetchPage()
        assertEquals(1, extractor.streamCount)
    }

    class ComingOfAge : BasePlaylistExtractorTest {
        @Test
        @Throws(ParsingException::class)
        override fun testThumbnails() {
            BandcampTestUtils.testImages(extractor!!.thumbnails)
        }

        @Test
        @Throws(ParsingException::class)
        override fun testBanners() {
            ExtractorAsserts.assertEmpty(extractor!!.banners)
        }

        @Test
        @Throws(ParsingException::class)
        fun testUploaderUrl() {
            assertTrue(extractor!!.uploaderUrl!!.contains("macbenson.bandcamp.com"))
        }

        @Test
        @Throws(ParsingException::class)
        override fun testUploaderName() {
            assertEquals("mac benson", extractor!!.uploaderName)
        }

        @Test
        @Throws(ParsingException::class)
        override fun testUploaderAvatars() {
            BandcampTestUtils.testImages(extractor!!.uploaderAvatars)
        }

        @Test
        @Throws(ParsingException::class)
        override fun testStreamCount() {
            assertEquals(5, extractor!!.streamCount)
        }

        @Test
        @Throws(ParsingException::class)
        fun testDescription() {
            val description = extractor!!.description
            Assertions.assertNotEquals(Description.EMPTY_DESCRIPTION, description)
            ExtractorAsserts.assertContains("Artwork by Shona Radcliffe", description.content) // about
            ExtractorAsserts.assertContains("All tracks written, produced and recorded by Mac Benson",
                    description.content) // credits
            ExtractorAsserts.assertContains("all rights reserved", description.content) // license
        }

        @Test
        @Throws(Exception::class)
        override fun testUploaderVerified() {
            assertFalse(extractor!!.isUploaderVerified)
        }

        @Test
        @Throws(IOException::class, ExtractionException::class)
        fun testInitialPage() {
            Assertions.assertNotNull(extractor!!.initialPage.getItems().get(0))
        }

        @Test
        override fun testServiceId() {
            Assertions.assertEquals(Bandcamp.serviceId, extractor!!.serviceId)
        }

        @Test
        @Throws(ParsingException::class)
        override fun testName() {
            assertEquals("Coming of Age", extractor!!.name)
        }

        @Test
        @Throws(Exception::class)
        override fun testId() {
            Assertions.assertEquals("https://macbenson.bandcamp.com/album/coming-of-age", extractor!!.id)
        }

        @Test
        @Throws(Exception::class)
        override fun testUrl() {
            Assertions.assertEquals("https://macbenson.bandcamp.com/album/coming-of-age", extractor!!.url)
        }

        @Test
        @Throws(Exception::class)
        override fun testOriginalUrl() {
            Assertions.assertEquals("https://macbenson.bandcamp.com/album/coming-of-age", extractor!!.originalUrl)
        }

        @Test
        @Throws(IOException::class, ExtractionException::class)
        fun testNextPageUrl() {
            Assertions.assertNull(extractor!!.getPage(extractor!!.initialPage!!.nextPage))
        }

        @Test
        @Throws(Exception::class)
        override fun testRelatedItems() {
            // DefaultTests.defaultTestRelatedItems(extractor);
            // Would fail because BandcampPlaylistStreamInfoItemExtractor.getUploaderName() returns an empty String
        }

        @Test
        @Throws(Exception::class)
        override fun testMoreRelatedItems() {
        }

        companion object {
            private var extractor: PlaylistExtractor? = null
            @BeforeAll
            @Throws(ExtractionException::class, IOException::class)
            fun setUp() {
                init(DownloaderTestImpl.Companion.getInstance())
                extractor = Bandcamp.getPlaylistExtractor("https://macbenson.bandcamp.com/album/coming-of-age")
                extractor!!.fetchPage()
            }
        }
    }

    companion object {
        @BeforeAll
        fun setUp() {
            init(DownloaderTestImpl.Companion.getInstance())
        }
    }
}
